File: TeamCode/src/main/java/org/firstinspires/ftc/teamcode/Tasks/Task.java
package org.firstinspires.ftc.teamcode.Tasks;

import org.firstinspires.ftc.teamcode.OpModes.Autonomous.AutoBase;

public abstract class Task {
    public boolean isFinished() {
        return state == State.FINISHED;
    }

    public boolean isRunning() {
        return state == State.RUNNING;
    }

    public enum State {
        DEFAULT,
        RUNNING,
        FINISHED,
    }

    public State state = State.DEFAULT;
    public AutoBase context;

    public void run() {};
    public void tick() {};

    public void start(AutoBase context) {
        this.context = context;
        state = State.RUNNING;
        run();
    }
}

File: TeamCode/src/main/java/org/firstinspires/ftc/teamcode/tasks/Task.kt
package org.firstinspires.ftc.teamcode.tasks

import com.qualcomm.robotcore.eventloop.opmode.LinearOpMode

enum class TaskState {
    RUNNING,
    FINISHED,
}

abstract class Task {
    var state = TaskState.RUNNING
    open fun tick() {}
    abstract fun run()
    lateinit var context: LinearOpMode
    fun start(ctx: LinearOpMode) {
        state = TaskState.RUNNING
        context = ctx
        run()
    }
    fun isRunning() = state == TaskState.RUNNING
    fun isFinished() = state == TaskState.FINISHED
}